File: Data Structures & Algorithms/redundant-connection/submission-4.js
class Solution {
    /**
     * @param {number[][]} edges
     * @return {number[]}
     */
    findRedundantConnection(edges) {
        let adj = new Map();

        for(let i = 1; i <= edges.length; i++){
            adj.set(i, []);
        }

        for(const [src, end] of edges){
            adj.get(src).push(end);
            adj.get(end).push(src);

            let seen = new Set();

            if(dfs(src, seen, -1)) return [src, end];
        }

        function dfs(start, seen, parent){
            if(seen.has(start)) return true;

            seen.add(start);

            for(const nei of adj.get(start)){
                if(nei === parent) continue;
                if(dfs(nei, seen, start)) return true;
            }

            return false;
        }
        return [];
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-4.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = {}
        res = 0

        for i in range(n):
            adj[i] = []

        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        seen = set()
        def dfs(node, parent):
             
            seen.add(node)

            for edge in adj[node]:
                if not edge == parent and not edge in seen:
                    dfs(edge, node)
        
        for i in range(n):
            if not i in seen:
                dfs(i, None)
                res += 1
        
        return res
            

            